halo2_proofs/src/plonk/circuit.rs

```diff
@@ -706,6 +792,51 @@ pub enum Expression<F> {
 }
 
 impl<F: Field> Expression<F> {
+    /// Make side effects
+    pub fn query_cells(&mut self, cells: &mut VirtualCells<'_, F>) {
```

Comment:
This should be decently parallelizable no? Can't you split all the `cells` into chunks and digest them in parallel? Maybe `Mutex`s add too much overhead and it isn't worth it. But might be worth a try in the future probably.

Reply:
This is just to make the query API easier to use, but it is a good idea to parallelize when processing `Vec<Expression<F>>` in `Gate`, or even aggregate all expressions in `ConstraintSystem` and parallelize them all. We can open an issue on that.